### ISSUE TITLE
Syntax highlighting

### ISSUE THREAD

**Issue by davidtheclark** [mapbox/jsxtreme-markdown#10]
Writing down some thoughts about syntax highlighting, after I thought about it a little too much today:

I've been waffling between [Prism](http://prismjs.com/) and [highlight.js](https://highlightjs.org/).

My impression is that Prism provides better support for web languages. It is used by MDN, Smashing Magazine, A List Apart, so it is no doubt very good with web languages. Looking at Prism and highlight.js's treatment of JS back-to-back shows [Prism is better at it](https://github.com/gatsbyjs/gatsby/issues/1077).

highlight.js, though, seems to be a little more popular; bigger, older; and a little more active.

Both are JS libraries that put some effort into being small and efficient for use in the browser —— which are not really concerns here. So we *could* consider other highlighters. I just don't know much about others that we should consider.

I learned that [npm uses](https://github.com/npm/marky-markdown) Atom editor's syntax highlighting engine ([blog post](http://blog.npmjs.org/post/109508231330/nicely-presented-markup)), [highlights](https://github.com/atom/highlights). One advantage of highlights is that it can read language grammars, not just regular expressions. In theory, we should get as much syntax highlighting control as the Atom editor. However, it doesn't look like there's a well maintained Atom grammar for Swift?!

@1ec5 do you have any sense whether using highlights would make it easier for us to get really great syntax highlighting for Swift and Objective-C? Do you have any sense whether Atom's support for these languages is already good enough?

**Comment by 1ec5** [comment]
> do you have any sense whether using highlights would make it easier for us to get really great syntax highlighting for Swift and Objective-C?

First off, I love Highlights’ logo. 🌈 Highlights’ Objective-C support is robust but a couple years outdated. It [lacks built-in support for Swift](https://github.com/atom/highlights/tree/master/deps/), but there is a [third-party highlighter](https://github.com/jnewc/atom-language-swift/). I haven’t tried it out, but the grammar appears to be ad-hoc rather than being based on [the official BNF grammar](https://developer.apple.com/library/content/documentation/Swift/Conceptual/Swift_Programming_Language/zzSummaryOfTheGrammar.html). This makes it more difficult to update the grammar for this year’s Swift 4 (multiline strings, key paths) and beyond.

By comparison, highlight.js and Prism are quite primitive when it comes to Objective-C support, since they rely on regular expressions alone. highlight.js has already been updated to last year’s Swift 3 (isagalaev/highlight.js#1532), whereas Prism is presumably [still on Swift 2](https://github.com/PrismJS/prism/blob/gh-pages/components/prism-swift.js).

For really great syntax highlighting of Swift code, there’s really no way around generating a proper AST using a parser. Anything less will inevitably result in either underhighlighting or overhighlighting, due to string interpolation, all the contextual keywords, and the opportunity for custom operators. Highlights does implement a parser, but the list of standard library symbols to highlight is hard-coded, incomplete, and perennially outdated. (This goes for Objective-C as well.)

[SourceKit](https://github.com/apple/swift/tree/master/tools/SourceKit) via [SourceKitten](https://github.com/jpsim/SourceKitten/) via [node-sourcekit](https://github.com/RLovelett/node-sourcekit/) would automatically get us proper syntax highlighting of Swift and Objective-C, including highlighting standard library symbols (which is consistent with how Xcode presents code). Note that node-sourcekit requires either a Linux or Mac environment.

In short, Atom’s support is better than the others, but it’s still jarring for an iOS developer to see. (It’s like opening a Swift file on its own in Xcode, independently of its project, and seeing nothing but the keywords highlighted.)

/ref https://github.com/mapbox/ios-sdk/issues/155#issuecomment-224138555

**Comment by davidtheclark** [comment]
I'm leaning towards highlights this morning, because it is powerful and extensible, without concern for running in the browser; and it's good to be on the same side as Atom and npm, both of which would provide extensive battle testing.

@1ec5: If I understand this correctly, we want a tool that:

- Uses node-sourcekit to create JSON syntax information: https://github.com/jpsim/SourceKitten/#syntax
- Writes HTML with `<span>`s and class names for each syntax type. Ideally, class names would correspond with [Atom syntax themes](http://flight-manual.atom.io/hacking-atom/sections/creating-a-theme/#creating-a-syntax-theme).

I wonder if we should try to do this as a contribution to https://github.com/jnewc/atom-language-swift/ or as a new tool.

**Comment by 1ec5** [comment]
> I wonder if we should try to do this as a contribution to https://github.com/jnewc/atom-language-swift/ or as a new tool.

I wonder what the project’s maintainers would think, but a dependency on SourceKit would have pros and cons. On the one hand, it would be more correct and maintainable. On the other hand, it would be less compatible, especially since Atom can run in environments where SourceKit cannot (e.g., Windows).

**Comment by davidtheclark** [comment]
Ooo, if SourceKit cannot run in Windows, that seems like a crushing blow for a generic Atom language package. Let's make our own thing.

**Comment by davidtheclark** [comment]
@1ec5: Just to confirm, before I start working on this ... you feel pretty confident that this will work on Linux (for automated builds)?

**Comment by 1ec5** [comment]
node-sourcekit [claims](https://github.com/RLovelett/node-sourcekit#support) to support Ubuntu 14.04. What that means in practice is that it’ll have access to the same Swift standard library and [Foundation](https://github.com/apple/swift-corelibs-foundation/) symbols, but it won’t have UIKit symbols. I think this will still be an improvement over the current situation, but maybe @captainbarbosa @jmkiley @friedbunny can chime in about whether it would be an issue that UIKit symbols are highlighted differently than Foundation symbols. There’s also the possibility of introducing hacks like defining our own stub UIKit that vends the symbols but doesn’t implement them.

**Comment by davidtheclark** [comment]
Let's move this discussion into our syntax-highlighing-specific repo. jsxtreme can take remark plugins — that's enough for it.
